central/networkpolicies/datastore/datastore_sac_test.go

```diff
@@ -0,0 +1,396 @@
+			scopeKey: testutils.Cluster1NamespaceBReadWriteCtx,
+		},
+		"read-write on matching cluster can count": {
+			scopeKey:      testutils.Cluster2ReadWriteCtx,
+			expectedCount: 1,
+		},
+		"read-write on matching cluster but wrong namespace cannot count": {
+			scopeKey: testutils.Cluster2NamespaceAReadWriteCtx,
+		},
+		"read-write on matching cluster and namespace can count": {
+			scopeKey:      testutils.Cluster2NamespaceBReadWriteCtx,
+			expectedCount: 1,
+		},
+		"read-write on matching cluster and at least one matching namespace can count": {
+			scopeKey:      testutils.Cluster2NamespacesABReadWriteCtx,
+			expectedCount: 1,
+		},
+	}
+
+	for name, c := range cases {
+		s.Run(name, func() {
+			ctx := s.testContexts[c.scopeKey]
+			count, err := s.datastore.CountMatchingNetworkPolicies(ctx, testconsts.Cluster2, testconsts.NamespaceB)
+			s.Equal(c.expectedCount, count)
+			s.Nil(err)
+		})
+	}
+}
+
+func (s *networkPolicySACSuite) TestUpsertNetworkPolicy() {
+	cases := map[string]struct {
+		scopeKey      string
+		expectedError bool
+	}{
+		"global read-only cannot upsert": {
+			scopeKey:      testutils.UnrestrictedReadCtx,
+			expectedError: true,
+		},
+		"global read-write can upsert": {
+			scopeKey: testutils.UnrestrictedReadWriteCtx,
+		},
+		"read-write on wrong cluster cannot upsert": {
+			scopeKey:      testutils.Cluster1ReadWriteCtx,
+			expectedError: true,
+		},
+		"read-write on wrong cluster and namespace cannot upsert": {
+			scopeKey:      testutils.Cluster1NamespaceAReadWriteCtx,
+			expectedError: true,
+		},
+		"read-write on wrong cluster and matching namespace cannot upsert": {
+			scopeKey:      testutils.Cluster1NamespaceBReadWriteCtx,
+			expectedError: true,
+		},
+		"read-write on matching cluster can upsert": {
+			scopeKey: testutils.Cluster2ReadWriteCtx,
+		},
+		"read-write on matching cluster but wrong namespace cannot upsert": {
+			scopeKey:      testutils.Cluster2NamespaceAReadWriteCtx,
+			expectedError: true,
+		},
+		"read-write on matching cluster and namespace can upsert": {
+			scopeKey: testutils.Cluster2NamespaceBReadWriteCtx,
+		},
+		"read-write on matching cluster and at least one matching namespace can upsert": {
+			scopeKey: testutils.Cluster2NamespacesABReadWriteCtx,
+		},
+	}
+
+	for name, c := range cases {
+		s.Run(name, func() {
+			unrestrictedCtx := s.testContexts[testutils.UnrestrictedReadWriteCtx]
+			ctx := s.testContexts[c.scopeKey]
+			policy := fixtures.GetScopedNetworkPolicy(uuid.NewV4().String(), testconsts.Cluster2, testconsts.NamespaceB)
+			err := s.datastore.UpsertNetworkPolicy(ctx, policy)
+			defer s.deleteNetworkPolicy(policy.GetId())
+			if c.expectedError {
+				s.ErrorIs(err, sac.ErrResourceAccessDenied)
+			} else {
+				s.NoError(err)
+				count, countErr := s.datastore.CountMatchingNetworkPolicies(unrestrictedCtx, testconsts.Cluster2, testconsts.NamespaceB)
+				s.Equal(1, count)
+				s.NoError(countErr)
```

Comment:
nit: I favor checking for the error first before asserting on the returned value.